cf/data/data.py

```diff
@@ -1034,78 +1133,202 @@ def __setitem__(self, indices, value):
+            self.roll(shift=shifts, axis=roll_axes, inplace=True)
 
         # Make sure that the units of value are the same as self
-        try:
-            value_units = value.Units
-        except AttributeError:
-            pass
-        else:
-            self_units = self.Units
-            if value_units.equivalent(self_units):
-                if value_units != self_units:
-                    value = value.copy()
-                    value.Units = self.Units
-            elif value_units and self_units:
-                raise ValueError(
-                    f"Can't assign values with units {value_units!r} "
-                    f"to data with units {self_units!r}"
-                )
-        # --- End: try
-
-        # Set the mask hardness, in case any previous operations have
-        # inadvertently changed it.
-        self._set_mask_hardness()
+        value = conform_units(value, self.Units)
 
         # Do the assignment
         dx = self._get_dask()
         dx[indices] = dask_compatible(value)
 
+        # Unroll any axes that were rolled to enable a cyclic
+        # assignment
         if roll:
-            # Unroll any axes that were rolled to enable a cyclic
-            # assignment
             shifts = [-shift for shift in shifts]
-            self.roll(axis=roll_axes, shift=shifts, inplace=True)
+            self.roll(shift=shifts, axis=roll_axes, inplace=True)
+
+        # Reset the mask hardness, otherwise it could be incorrect in
+        # the case that a chunk that was not a masked array is
+        # assigned missing values.
+        self._reset_mask_hardness()
 
         return
 
+    # ----------------------------------------------------------------
+    # Indexing behaviour attributes
+    # ----------------------------------------------------------------
+    @property
+    @daskified(1)
+    def __orthogonal_indexing__(self):
+        """Flag to indicate that orthogonal indexing is supported.
+
+        Always True, indicating that 'orthogonal indexing' is
+        applied. This means that when indices are 1-d arrays or lists
+        then they subspace along each dimension independently. This
+        behaviour is similar to Fortran, but different to `numpy`.
+
+        .. versionadded:: TODODASK
+
+        .. seealso:: `__keepdims_indexing__`, `__getitem__`,
+                     `__setitem__`,
+                     `netCDF4.Variable.__orthogonal_indexing__`
+
+        **Examples**
+
+        >>> d = cf.Data([[1, 2, 3],
+        ...              [4, 5, 6]])
+        >>> e = d[[0], [0, 2]]
+        >>> e.shape
+        (1, 2)
+        >>> print(e.array)
+        [[1 3]]
+        >>> e = d[[0, 1], [0, 2]]
+        >>> e.shape
+        (2, 2)
+        >>> print(e.array)
+        [[1 3]
+         [4 6]]
+
+        """
+        return True
+
+    @property
+    @daskified(1)
+    def __keepdims_indexing__(self):
+        """Flag to indicate whether orthogonal indexing is supported.
```

Comment:
This summary line needs updating I imagine since it seems to be identical to the `__orthogonal_indexing__` one, probably from copy & paste?

Reply:
Indeed ...

```suggestion
        """Flag to indicate whether dimensions indexed with integers are kept.
```